prepared_cache.go

```diff
@@ -1,37 +1,101 @@
 package gocql
 
 import (
-	"github.com/gocql/gocql/internal/lru"
 	"sync"
+	"sync/atomic"
+
+	"github.com/gocql/gocql/internal/lru"
 )
 
 const defaultMaxPreparedStmts = 1000
 
-// preparedLRU is the prepared statement cache
-type preparedLRU struct {
-	mu  sync.Mutex
-	lru *lru.Cache
+type preparedCache struct {
+	hosts     atomic.Value // map[string]*preparedLRU
+	mu        sync.Mutex
+	maxStmnts int
+}
+
+func (p *preparedCache) load() map[string]*preparedLRU {
+	v, _ := p.hosts.Load().(map[string]*preparedLRU)
+	return v
 }
 
-// Max adjusts the maximum size of the cache and cleans up the oldest records if
-// the new max is lower than the previous value. Not concurrency safe.
-func (p *preparedLRU) max(max int) {
+func (p *preparedCache) clear() {
 	p.mu.Lock()
-	defer p.mu.Unlock()
+	p.hosts.Store(make(map[string]*preparedLRU))
+	p.mu.Unlock()
+}
+
+func (p *preparedCache) forHost(addr string) *preparedLRU {
+	m := p.load()
+	lru, ok := m[addr]
+	if !ok {
+		p.mu.Lock()
+		m := p.load()
+		lru, ok = m[addr]
+		if !ok {
+			lru = p.addHost(m, addr)
+		}
+		p.mu.Unlock()
+	}
 
-	for p.lru.Len() > max {
-		p.lru.RemoveOldest()
+	return lru
+}
+
+func (p *preparedCache) addHost(current map[string]*preparedLRU, addr string) *preparedLRU {
```

Comment:
This method is not synchronized. It is called from the event notification as well directly. If that is perfectly serialized then we are just fine but otherwise this might need protections.